Inductive invariant validation.

lemmy opened the pull request:
* https://github.com/apalache-mc/apalache/blob/main/docs/src/apalache/known-issues.md#integer-ranges-with-non-constant-bounds
* https://github.com/apalache-mc/apalache/issues/425
* https://github.com/apalache-mc/apalache/pull/1431

Github issue #5
https://github.com/heidihoward/pbft-tlaplus/issues/5

[Apalache]

lemmy committed:
Inductive invariant validation with Apalache.

* https://github.com/apalache-mc/apalache/blob/main/docs/src/apalache/known-issues.md#integer-ranges-with-non-constant-bounds
* https://github.com/apalache-mc/apalache/issues/425
* https://github.com/apalache-mc/apalache/pull/1431

Github issue #5
https://github.com/heidihoward/pbft-tlaplus/issues/5

[Apalache]

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Build main/HEAD of Apalache until new official Apalache builds are available.

https://github.com/apalache-mc/apalache/issues/2951

[Build]

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Remove bogus `www.` from url that causes a https cert error.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Consume newest Apalache release.

https://github.com/apalache-mc/apalache/issues/2951

[Apalache]

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy commented on pbft.tla:
```diff
@@ -591,8 +591,11 @@ GetDigest(ppms, sn) ==
+        maxs == Max0({ppm.n: ppm \in ppms}) 
+        \* Apalache does not yet support integer ranges with non-constant bounds:
+        \* https://github.com/apalache-mc/apalache/blob/main/docs/src/apalache/known-issues.md#integer-ranges-with-non-constant-bounds
+        apa == { j \in Views : mins+1 <= j /\ j <= maxs} IN
```
@heidihoward Could you please confirm that it is correct to quantify over `Views`?

heidihoward replied:
it should be quantified over `SeqNums` instead

lemmy committed:
Install TLC nightly build to consume https://github.com/tlaplus/tlaplus/commit/6a1ea006b4561f0978b5e706ebc7f81c0679b1e2

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Switch to TLC's `generate` mode to prevent it from generating an action's complete set of successor states.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Leave note why Tstamp and Views have to be redefined to be finite sets.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Remove unused ConstInit.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
TLAPM does not (yet) support bound tuples (https://github.com/tlaplus/tlapm/pull/140)

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Ignore TLAPS fingerprint cache.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Rewrite TypeOK into monstrous universal quantification that avoids constraint explosion.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Decrease number of generates values to prevent Apalache from exceeding its heap memory.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
mlogs is defined for all replicas (outside of TLA+, one would call it a total function). Without this constraint, Apalache!Gen may generate a function that is not defined for all replicas

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Remove incomplete leftover.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Partial inductive invariant checking with Apalache.

Checks  SafetyInv  instead of  Inv  because TypeOK's  subsets of subsets of subsets makes Apalache's encoding explode.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
No surprise that SafetyInv is violated if *all* nodes are Byzantine.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Push Apalache!Gen into dedicated .tla file to prevent TLC from failing due to missing Apalache.tla import.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Assume that the number of byzantine replicas is less than what the protocol can tolerate.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Generate, simulate, validate, and verify with TLC and Apalache.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Mitigate state-space explosion of two byzantine actions to prevent TLC from attempting to generate sets that are too large.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Do not parse *all* specs because Apalache module is not there.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Finish in reasonable time of approx 10 minutes.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Bug: Range over constant SeqNums instead of Views.

https://github.com/heidihoward/pbft-tlaplus/pull/6/commits/48316b6f0f812ea03fabde67f8a31f7fd9fbb8fe#r1745808451
Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

lemmy committed:
Simulation starting at GenInit predicate with Apalache.

Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>
